hawc/apps/animalv2/views.py

```diff
@@ -0,0 +1,343 @@
+    """
+
+    def get_success_url(self):
+        return self.object.study.get_absolute_url()
+
+
+class ExperimentViewSet(HtmxViewSet):
+    actions = {"read", "update"}
+    parent_model = Study
+    model = models.Experiment
+    form_fragment = "animalv2/fragments/_experiment_edit.html"
+    detail_fragment = "animalv2/fragments/_experiment_table.html"
+
+    @action(permission=can_view)
+    def read(self, request: HttpRequest, *args, **kwargs):
+        return render(request, self.detail_fragment, self.get_context_data())
+
+    @action(methods=("get", "post"), permission=can_edit)
+    def update(self, request: HttpRequest, *args, **kwargs):
+        template = self.form_fragment
+        data = request.POST if request.method == "POST" else None
+        form = forms.ExperimentForm(data=data, instance=request.item.object)
+        if request.method == "POST" and form.is_valid():
+            self.perform_update(request.item, form)
+            template = self.detail_fragment
+        context = self.get_context_data(form=form)
+        return render(request, template, context)
+
+
+class ExperimentChildViewSet(HtmxViewSet):
+    actions = {"create", "read", "update", "delete", "clone"}
+    parent_model = models.Experiment
+    model = None  # required
+    form_class: str  # required
+    form_fragment = "common/fragments/_object_edit_row.html"
+    detail_fragment: str  # required
+    formset_configurations = []
+
+    @action(permission=can_view)
+    def read(self, request: HttpRequest, *args, **kwargs):
+        return render(request, self.detail_fragment, self.get_context_data())
+
+    @action(methods=("get", "post"), permission=can_edit)
+    def create(self, request: HttpRequest, *args, **kwargs):
+        template = self.form_fragment
+        formsets = []
+        formsets_valid_if_present = True
+        if request.method == "POST":
+            # make a copy; if we do any is_valid modifying of the data we need this...
+            request.POST = request.POST.copy()
+
+            form = self.form_class(request.POST, parent=request.item.parent)
+
+            for formset_config in self.formset_configurations:
+                formset = modelformset_factory(
+                    formset_config.model_class,
+                    form=formset_config.form_class,
+                    can_delete=True,
+                )(request.POST, prefix=formset_config.form_prefix)
+                formsets.append(formset)
+
+                if not formset.is_valid():
+                    formsets_valid_if_present = False
+
+            if form.is_valid() and formsets_valid_if_present:
+                self.perform_create(request.item, form, formsets)
+                template = self.detail_fragment
+        else:
+            form = self.form_class(parent=request.item.parent)
+        context = self.get_context_data(form=form, formsets=formsets)
+        return render(request, template, context)
+
+    # TODO - update the update method
```

Comment:
elaborate on this comment or remove? i'm not sure what the action item is.

Reply:
I have zero recollection what this refers to anymore. Removed.